### PR TITLE
Update rf

### PR DESCRIPTION
**Pull request by guabao**
# For updates and changes
Changes:
1. Add experimental training pipeline in client which supports different network topologies
2. Update random forest demo so that the demo fits the experimental training pipeline

### CODE REVIEW

**Review by cyqclark — APPROVED** [review]
LGTM

**Review by fedlearnJDT — COMMENTED** [review]
LGTM

File: demos/random_forest/new_coordinator.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
+# Copyright 2021 Fedlearn authors.
```
Comment (cyqclark) [inline demos/random_forest/new_coordinator.py:1]: Add a todo of refactoring the whole file once update is done, otherwise it will be confusing to have multiple files with similar names.

File: demos/random_forest/client.py [hunk 2]
```diff
@@ -687,5 +690,20 @@ def encrypt(x):
     parameter = config.parameter
     client = RandomForestClient(client_info, parameter, dataset, remote=True)
 
-    serve(client)
+    # set active client
+    if config.active_index == args.index:
+        coordinator_info = MachineInfo(ip=ip, port=port,
+                                       token=config.coordinator_ip_and_port)
+        client_infos = []
+        for ci in config.client_ip_and_port:
```
Comment (cyqclark) [inline demos/random_forest/client.py:698]: better not using ci as a variable name for better readability

File: demos/random_forest/client.py [hunk 3]
```diff
@@ -663,7 +666,7 @@ def encrypt(x):
     parser = argparse.ArgumentParser()
     parser.add_argument('-I', '--index', type=int, required=True, help='index of client')
     parser.add_argument('-C', '--python_config_file', type=str, required=True, help='python config file')
-
+
```
Comment (cyqclark) [inline demos/random_forest/client.py:669]: unnecessary change

File: demos/random_forest/client.py [hunk 4]
```diff
@@ -45,6 +45,8 @@
 from core.entity.common.machineinfo import MachineInfo
 from core.entity.common.message import RequestMessage, ResponseMessage
 from core.client.client import Client
+#from new_client import Client
```
Comment (cyqclark) [inline demos/random_forest/client.py:48]: Remove if not needed, sort by alphabetical order

File: core/client/client.py [hunk 5]
```diff
@@ -119,3 +122,81 @@ def comm(self, grpc_request: ReqResMessage, context) -> ReqResMessage:
...
+        if (client_info.ip == self._client_info.ip) and (
+            client_info.port == self._client_info.port):
+            response = self._exp_call_local_client(request)
+        else:
+            response = send_request(request)
+        return response
+
+    def _exp_call_grpc_client(self, requests, is_parallel=True):
+        responses = {}
+        """
```
Comment (cyqclark) [inline core/client/client.py:151]: If this is temporary, add a todo

File: core/client/client.py [hunk 6]
```diff
@@ -15,6 +15,7 @@
 from core.entity.common.message import RequestMessage, ResponseMessage
 from core.grpc_comm.grpc_converter import grpc_msg_to_common_msg, common_msg_to_grpc_msg
 from core.proto.transmission_pb2 import ReqResMessage
+from core.grpc_comm.grpc_client import send_request
```
Comment (cyqclark) [inline core/client/client.py:18]: Sort by alphabetical orders
